internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubMessages.java

```diff
@@ -0,0 +1,108 @@
+package com.linkedin.venice.pubsub;
+
+import com.linkedin.venice.pubsub.api.PubSubMessage;
+import com.linkedin.venice.pubsub.api.PubSubTopic;
+import com.linkedin.venice.pubsub.api.PubSubTopicPartition;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.kafka.common.utils.AbstractIterator;
+
+
+public class PubSubMessages<K, V, OFFSET> implements Iterable<PubSubMessage<K, V, OFFSET>> {
```

Comment:
Is this necessary? I'm not a fan of Kafka's `ConsumerRecords` so not enthusiastic about copying it here. Everything is inefficient in this class... Lots of useless iterations, collection copies, immutable wrappers... We have no choice to use it when interacting with Kafka, but it seems unfortunate to pay all of these costs twice (once for Kafka, once for our own copy...).

I would prefer that we just return a `Map<PubSubTopicPartition, List<PubSubMessage<K, V, OFFSET>>>`, and not this wrapper object.

Reply:
Initially I would like to use `Map<PubSubTopicPartition, List<PubSubMessage<K, V, OFFSET>>>`, but then I found that we have many places using the iterator of these 2-layered (e.g. `KafkaInputRecordReader`, `AdminConsumptionTask`). I think this is up to which way we want to go with the interface of `PubSubConsumer#poll()`. If we go with `Map<PubSubTopicPartition, List<PubSubMessage<K, V, OFFSET>>>`, we should have some logic either inside `ApacheKafkaConsumer` or outside to flatter this two layered structure.

Reply:
I would like us to have a bias towards optimizing the hot path, which I think would benefit from returning just an unwrapped map.

Regarding usability of the non-hot paths, here are a few alternatives:

1. Create some `Utils::iterateOnMapOfLists(Map<K, List<V>> map)` which returns an `Iterable<V>`. The callers can then do something like `for (V value: Utils.iterateOnMapOfLists(consumer.poll()))`...
2. Have two poll APIs in the consumer, one which returns `Map<K, List<V>>` and the other which returns `Iterable<V>`.
3. Create a special `MapOfLists<K, V>` interface which extends `Map<K, List<V>>` and which will have a function for returning an `Iterable<V>`.

Does any of these option appear ok to you?

Reply:
I would like to choose 1 to have `Utils::iterateOnMapOfLists(Map<K, List<V>> map)` to keep the interface of `PubSubConsumer` simpler.